src/test/crush/crush.cc

```diff
@@ -388,6 +388,309 @@ INSTANTIATE_TEST_SUITE_P(
   ::testing::Values(RuleType(true), RuleType(false)),
   testing::PrintToStringParamName());
 
+class FirstnTest : public ::testing::TestWithParam<RuleType>
+{
+public:
+  void SetUp() final
+  {
+    CephInitParameters params(CEPH_ENTITY_TYPE_CLIENT);
+    cct = common_preinit(params, CODE_ENVIRONMENT_UTILITY,
+			 CINIT_FLAG_NO_DEFAULT_CONFIG_FILE);
+  }
+  void TearDown() final
+  {
+    cct->put();
+    cct = nullptr;
+  }
+
+  std::unique_ptr<CrushWrapper> build_firstn_map(
+    CephContext *cct, int num_rack, int num_host, int num_osd)
+  {
+    std::unique_ptr<CrushWrapper> c(new CrushWrapper);
+    c->create();
+    c->set_tunables_optimal();
+
+    c->set_type_name(5, "root");
+    c->set_type_name(4, "row");
+    c->set_type_name(3, "rack");
+    c->set_type_name(2, "chasis");
+    c->set_type_name(1, "host");
+    c->set_type_name(0, "osd");
+
+    int rootno;
+    c->add_bucket(0, CRUSH_BUCKET_STRAW, CRUSH_HASH_RJENKINS1,
+		  5, 0, NULL, NULL, &rootno);
```

Comment:
nit: `nullptr` as this is a new code which, partially, already uses this variant.